### PR TITLE
remove Cylc 7 compatibility mode

### LINKED ISSUE #6849: drop compatibility mode
Cylc 7 is now long dead, it's time to schedule the removal of compatibility mode.

We should pick a future Cylc 8 version to withdraw support (e.g. 8.7.0 or 8.8.0) and include this in the deprecation notice. We should also announce this on Discourse.

We should also 

See also (would make sense to do these at the same time):
* https://github.com/cylc/cylc-rose/issues/374
* https://github.com/cylc/cylc-flow/issues/6850

### LINKED ISSUE #6850: jinja2: bump version
We are running with an old version of Jinja2:

https://github.com/cylc/cylc-flow/blob/7bf22339b297a7a7a93b17daf6332fc6cd2c1597/setup.cfg#L70-L71

Jinja2 introduced a bunch of deprecations in 3.0.x for features which is subsequently removed in 3.1.x. To give our users a chance, we pinned to 3.0.x so they could see these deprecation warnings and respond to them before we ripped out this compatibility altogether.

It's been long enough now and compatibility mode as whole should be removed before too long - https://github.com/cylc/cylc-flow/issues/6849 - It's time to bump the version!

See also (would make sense to do these at the same time):

* https://github.com/cylc/cylc-rose/issues/374 
* https://github.com/cylc/cylc-flow/issues/6849

### PR DESCRIPTION
**Pull request by oliver-sanders**
A set of changes to remove Cylc 7 compatibility mode, issues:

* #6849 
* #6850
* https://github.com/cylc/cylc-rose/issues/374

Sibling PRs:

* https://github.com/cylc/cylc-rose/pull/429
* https://github.com/cylc/cylc-doc/pull/939

Summary:

* remove cylc 7 compatibility mode:
  * Closes https://github.com/cylc/cylc-flow/issues/6849
  * Remove Cylc 7 compatibility mode.
  * Remove support for validating, (re)installing or (re)starting `suite.rc` workflows.
  * Keep support for CLI commands against workflows running under older versions of Cylc 8.
  * Remove legacy functional tests.
  * Add integration tests (to ensure `suite.rc` workflows cannot be started).
* setup: unpin Jinja2:
  * Closes https://github.com/cylc/cylc-flow/issues/6850
  * Remove the upper pin on Jinja2 allowing the use of newer versions.
* suite.rc: schedule removal of remaining support
  * Cylc 8 is still `suite.rc` "aware" for now, it's needed for GUI/Tui/CLI.
  * Following the logic laid out in https://github.com/cylc/cylc-flow/pull/7274, this support would be removed in Cylc 8.10.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [ ] Does not contain off-topic changes - off topic changes:
  - jinja2: back support removed interfaces (requested)
  - config: remove redundant parser argument (requested in review)
  - scan: add test for get_workflow_title (shortfall spotted in review to ensure back-compat not broken)
  - parser arguments standardised (unrelated, but a pain for testing)
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/939.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/scripts/validate.py [hunk 1]
```diff
@@ -137,11 +137,11 @@ def get_option_parser():
 
 @cli_function(get_option_parser)
 def main(parser: COP, options: 'Values', workflow_id: str) -> None:
-    asyncio.run(run(parser, options, workflow_id))
+    asyncio.run(run(options, workflow_id))
 
 
 async def run(
-    parser: COP, options: 'Values', workflow_id: str
+    options: 'Values', workflow_id: str
```
Comment (oliver-sanders) [inline cylc/flow/scripts/validate.py:145]: *Off-topic change:* A couple of scripts were passing through the unused `parser` argument which was creating hassle for tests. These have been removed so all commands have the same basic interface.

File: cylc/flow/cfgspec/workflow.py [hunk 2]
```diff
@@ -165,10 +164,8 @@ def get_script_common_text(this: str, example: Optional[str] = None):
 
         .. versionchanged:: 8.0.0
 
-           The configuration file was previously named ``suite.rc``, but that
-           name is now deprecated.
-           The ``suite.rc`` file name now activates :ref:`cylc_7_compat_mode`.
-           Rename to ``flow.cylc`` to turn off compatibility mode.
+           The configuration file was rename from ``suite.rc`` to
```
Comment (hjoliver) [inline cylc/flow/cfgspec/workflow.py]:
```suggestion
           The configuration file was renamed from ``suite.rc`` to
```

**Review by hjoliver — APPROVED** [review]
Looks good, nice tidy up.

File: changes.d/6849.break.md [hunk 3]
```diff
@@ -0,0 +1 @@
+Cylc 7 compatibility mode has been removed, any workflows defined by a `suite.rc` file must now be upgraded to `flow.cylc` to run with the latest version of Cylc.
```
Comment (MetRonnie) [inline changes.d/6849.break.md]: (Minor)
```suggestion
Cylc 7 compatibility mode has been removed. Any workflows defined by a `suite.rc` file must now be renamed to `flow.cylc` to run with this version of Cylc.
```

File: changes.d/6850.break.md [hunk 4]
```diff
@@ -0,0 +1 @@
+The version of Jinja2 has been upgraded from 3.0 to 3.1, the version at which deprecated language features were removed.
```
Comment (MetRonnie) [inline changes.d/6850.break.md]:
```suggestion
The version of Jinja2 has been upgraded from 3.0 to 3.1, removing deprecated language features. See https://jinja.palletsprojects.com/en/stable/changes/#version-3-1-0 for details.
```

File: setup.cfg [hunk 5]
```diff
@@ -62,8 +62,7 @@ install_requires =
     colorama>=0.4,<1
     graphql-core>=3.2,<3.3
     graphene>=3.4.0,<3.5
-    # Note: can't pin jinja2 any higher than this until we give up on Cylc 7 back-compat
-    jinja2==3.0.*
+    jinja2>=3.0
```
Comment (MetRonnie) [inline setup.cfg]: 1. Needs updating here too: https://github.com/cylc/cylc-flow/blob/72669b79f8b0d1d4ba08a89ec2b4c79967021e15/conda-environment.yml#L10-L11

1. This comment: https://github.com/cylc/cylc-flow/issues/4702#issuecomment-1127407438 is the only explanation I can find for why we couldn't use 3.1 before this.
    > Note we must move to Jinja2 `3.0.*` not `3.*` initially as Jinja2 deprecated a bunch of things in `3.0.0` (with warnings) and removed these interfaces at `3.1.0`.
    > 
    > Moving to `3.1.0` would probably kill Cylc 7/8 interoperability for workflows which use Jinja2 globals/filters or `import jinja2` via `{% import` so we will probably have to stick with `3.0.*` until we are ready to cut off Cylc 7 interoperability.
    > 
    > Note we will want to advise users who wish to maintain Cylc 7 interoperability _not_ to take action on Jinja2 warnings until then.

    So does that mean users who had stuck with back-compat mode and now rename to `flow.cylc` and run in 8.7.0 might find it breaks without having seen any warnings? (I think not, because I can't find any suppression of Jinja2 warnings in the code - though I would like to test that warnings were coming through at all)

1. Do we want to upper pin to guard against being bitten by breaking changes in future?
    ```suggestion
        jinja2>=3.0,<3.2
    ```

File: cylc/flow/flags.py [hunk 6]
```diff
@@ -15,10 +15,14 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-"""Some global flags used in cylc"""
+"""Some global flags used in cylc.
+
+DEPRECATED:
+    We have been gradually whittling down the number of flags stored here.
+
+    Please do not add any further flags, to this module.
```
Comment (MetRonnie) [inline cylc/flow/flags.py:24]: (What is the better alternative?)

File: cylc/flow/task_pool.py [hunk 7]
```diff
@@ -381,17 +379,7 @@ def compute_runahead(self, force=False) -> bool:
...
-                    cylc.flow.flags.cylc7_back_compat and
-                    all(
-                        itask.state(TASK_STATUS_FAILED)
-                        for itask in itasks
-                    )
-                ):
-                    continue
+            for point in sorted(self.get_tasks_by_point()):
                 base_point = point
                 break
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py]:
```suggestion
            base_point = min(self.active_tasks.keys())
```

File: cylc/flow/workflow_files.py [hunk 8]
```diff
@@ -355,15 +361,16 @@ class ContactFileFields:
...
-NO_FLOW_FILE_MSG = (
+NO_FLOW_FILE_MSG_DEPR = (
     f"No {WorkflowFiles.FLOW_FILE} or {WorkflowFiles.SUITE_RC} "
     "in {}"
 )
+NO_FLOW_FILE_MSG = sstrip(f'''
+    No {WorkflowFiles.FLOW_FILE} in {{}}
+
+    Support for {WorkflowFiles.SUITE_RC} files was removed in Cylc 8.7.0:
+    https://cylc.github.io/cylc-doc/stable/html/7-to-8/major-changes/compatibility-mode.html
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py]: I think this part should only be shown if `suite.rc` is present, otherwise it is off-topic and might be confusing

File: cylc/flow/config.py [hunk 9]
```diff
@@ -295,7 +294,6 @@ def __init__(
                 See https://github.com/cylc/cylc-rose/issues/319
```
Comment (MetRonnie) [inline cylc/flow/config.py]: Also should remove the `force_compat_mode` arg?

File: cylc/flow/workflow_files.py [hunk 10]
```diff
@@ -1074,27 +1053,46 @@ def is_forbidden(flow_file: Path) -> bool:
     return False
 
 
-def check_flow_file(path: Union[Path, str]) -> Path:
+def check_flow_file(
+    path: Union[Path, str],
+    allow_suite_rc: bool = True,
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:1005]: Why not `False` by default?

File: tests/functional/optional-outputs/08-finish-fail-c7-c8.t
Comment (MetRonnie) [inline tests/functional/optional-outputs/08-finish-fail-c7-c8.t:1]: I think this test is still valid for the time being, as long as the upgrader is still present? (Would just do with permanent rename of `./finish-fail-c7/suite.rc` to `./finish-fail-c7/flow.cylc`)

File: setup.cfg (hunk 5 above)
Comment (MetRonnie) [reply setup.cfg]: Checked the deprecation warnings with the current version of Cylc (bottom warning - with a deliberate use of a deprecated import in a custom Jinja2 filter):
```console
$ export PYTHONWARNINGS=always
$ cylc config j2
.../lib/python3.12/site-packages/metomi/isodatetime/parsers.py:22: DeprecationWarning: module 'sre_constants' is deprecated
  import sre_constants

<template>:23: ResourceWarning: unclosed file <_io.TextIOWrapper name=6 mode='r' encoding='UTF-8'>
ResourceWarning: Enable tracemalloc to get the object allocation traceback

~/cylc-run/j2/Jinja2Filters/exclaim.py:6: DeprecationWarning: 'jinja2.escape' is deprecated and will be removed in Jinja 3.1. Import 'markupsafe.escape' instead.
  return f"{escape(val)}{escape(suffix)}"
```
Although I don't expect any users will have tried running with `PYTHONWARNINGS` set

File: tests/functional/cylc-install/02-failures.t
Comment (MetRonnie) [inline tests/functional/cylc-install/02-failures.t:1]: The code for this scenario could do with updating. At the least it should say "remove suite.rc" instead of "remove one".

Edit: also this part of the test duplicates `tests/unit/test_workflow_files.py::test_check_flow_file` 

https://github.com/cylc/cylc-flow/blob/72669b79f8b0d1d4ba08a89ec2b4c79967021e15/tests/functional/cylc-install/02-failures.t#L65-L76

File: tests/functional/cylc-play/07-timezones-compat.t
Comment (MetRonnie) [inline tests/functional/cylc-play/07-timezones-compat.t:1]: The second half of this test is still somewhat valid in that it is testing the log timestamps are in the local time zone when not in back-compat mode. However we don't need a functional test for this IMO

File: tests/functional/xtriggers/04-suite_state.t [hunk 11]
```diff
@@ -1,50 +0,0 @@
...
-[runtime]
-    [[foo]]
-__FLOW_CONFIG__
-
-msg='WARNING - The suite_state xtrigger is deprecated'
-
-TEST_NAME="${TEST_NAME_BASE}-val"
-run_ok "$TEST_NAME" cylc validate "$WORKFLOW_NAME"
-
-grep_ok "$msg" "${TEST_NAME}.stderr"
```
Comment (MetRonnie) [inline tests/functional/xtriggers/04-suite_state.t:43]: This first part of the test is still valid?

File: tests/integration/scripts/test_validate.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    ):
+        await validate(ValidateOptions(), str(tmp_path))
+
+    # flow.cylc present
+    (tmp_path / WorkflowFiles.SUITE_RC).unlink()
+    with pytest.raises(WorkflowConfigError):
+        await validate(ValidateOptions(), str(tmp_path))
+
+
+async def test_opposite_outputs(flow, validate):
```
Comment (MetRonnie) [inline tests/integration/scripts/test_validate.py]: Move to `tests/integration/validate/test_outputs.py`?

File: tests/integration/test_task_pool.py [hunk 13]
```diff
@@ -1748,10 +1740,7 @@ async def test_compute_runahead(
             TASK_STATUS_FAILED
         )
         schd.pool.compute_runahead(force=True)
-        if compat_mode == 'compat-mode':
-            assert int(str(schd.pool.runahead_limit_point)) == 5
-        else:
-            assert int(str(schd.pool.runahead_limit_point)) == 4  # no change
```
Comment (MetRonnie) [inline tests/integration/test_task_pool.py:1756]: Couple of now-outdated comments above
```diff
diff --git a/tests/integration/test_task_pool.py b/tests/integration/test_task_pool.py
index ecff8f565..4ab37b649 100644
--- a/tests/integration/test_task_pool.py
+++ b/tests/integration/test_task_pool.py
@@ -1728,14 +1728,12 @@ async def test_compute_runahead(
 
         # In Cylc 8 all incomplete tasks hold back runahead.
 
-        # In Cylc 7, submit-failed tasks hold back runahead..
         schd.pool.get_task(point('0001'), 'a').state.reset(
             TASK_STATUS_SUBMIT_FAILED
         )
         schd.pool.compute_runahead(force=True)
-        assert int(str(schd.pool.runahead_limit_point)) == 4
+        assert int(str(schd.pool.runahead_limit_point)) == 4  # no change
 
-        # ... but failed ones don't. Go figure.
         schd.pool.get_task(point('0001'), 'a').state.reset(
             TASK_STATUS_FAILED
         )
```

File: tests/unit/test_config.py [hunk 14]
```diff
@@ -1784,8 +1690,6 @@ def test_upg_wflow_event_names(back_compat, tmp_flow_config, log_filter):
     cfg = WorkflowConfig('foo', str(flow_file), ValidateOptions())
     for item in ('handler events', 'mail events'):
         assert cfg.cfg['scheduler']['events'][item] == expected
-    if back_compat:
-        assert not log_filter(logging.WARNING)
     else:
         assert log_filter(
             logging.WARNING,
```
Comment (MetRonnie) [inline tests/unit/test_config.py]:
```diff
diff --git a/tests/unit/test_config.py b/tests/unit/test_config.py
index 2e75c5afd..e695d7a20 100644
--- a/tests/unit/test_config.py
+++ b/tests/unit/test_config.py
@@ -1691,9 +1691,8 @@ def test_upg_wflow_event_names(tmp_flow_config, log_filter):
     for item in ('handler events', 'mail events'):
         assert cfg.cfg['scheduler']['events'][item] == expected
-    else:
-        assert log_filter(
-            logging.WARNING,
-            'Deprecated config items were automatically upgraded',
-        )
+    assert log_filter(
+        logging.WARNING,
+        'Deprecated config items were automatically upgraded',
+    )
 
 
```

File: cylc/flow/flags.py (hunk 6 above)
Comment (oliver-sanders) [reply cylc/flow/flags.py:24]: Local variables (as opposed to global ones).

Comment (hjoliver) [reply cylc/flow/flags.py:24]: For truly globally-applicable flags, global variables are still (at least arguably) appropriate, no?  (vs having to pass them through hundreds of argument lists...). Or is there always a local (ish) alternative? Like maybe our verbosity flag should belong to the logging module...

Comment (oliver-sanders) [reply cylc/flow/flags.py:24]: > For truly globally-applicable flags, global variables are still (at least arguably) appropriate, no

Clearly going with no!

The only valid global flag we have left is verbosity, which, for the scheduler at least could be easily stored on the `Scheduler` object.

Note this sort of thing is a right pain in tests (where we may be running multiple schedulers in the same Python instance) and multi-threadding (e.g, ZMQ) / multiprocessing (e.g, future xtriggers, etc) use cases.

E.G, remember the old Cylc 7 bug we had where grouping a family in one GUI also grouped it in another.

But also note that `lru_cache` presents very similar problems :(

Comment (hjoliver) [reply cylc/flow/flags.py:24]: > The only valid global flag we have left is verbosity, which, for the scheduler at least could be easily stored on the Scheduler object.

Well that one was my example 😁 and I didn't list any others (just thought in principle there could be!) - and it (verbosity) doesn't apply only to the scheduler so (at least arguably) it is still is appropriate as a global flag.

Ain't gonna argue that inappropriate use of globals is bad.

File: setup.cfg (hunk 5 above)
Comment (MetRonnie) [reply setup.cfg]: Warnings now introduced in https://github.com/cylc/cylc-flow/pull/7365

File: tests/functional/cylc-play/07-timezones-compat.t
Comment (oliver-sanders) [reply tests/functional/cylc-play/07-timezones-compat.t:1]: I'll restore it for now to keep this PR focused.

File: tests/functional/optional-outputs/08-finish-fail-c7-c8.t
Comment (oliver-sanders) [reply tests/functional/optional-outputs/08-finish-fail-c7-c8.t:1]: Good spot, I hadn't twigged the graph-deprecation nature of this test. Have removed the c7 part, renamed the `suite.rc`, removed the c7/8 part of the test name and amened comments.

File: cylc/flow/workflow_files.py (hunk 10 above)
Comment (oliver-sanders) [reply cylc/flow/workflow_files.py:1005]: Basically because 7/8 uses need `allow_suite_rc = True`.

Supporting the `suite.rc` file is still the "default" for most uses, e.g, we don't want a `cylc stop` command to fail because the workflow uses a `suite.rc` file. We're only removing support in targetted situations so this seemed safer.

File: cylc/flow/task_pool.py (hunk 7 above)
Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: Not quite, because `min([])` gives `ValueError` - hence `if base_point is None` on the line below.

File: cylc/flow/flags.py (hunk 6 above)
Comment (oliver-sanders) [reply cylc/flow/flags.py:24]: Yup, just trying to avoid accumulating more!

File: tests/unit/test_config.py (hunk 14 above)
Comment (oliver-sanders) [reply tests/unit/test_config.py]: Done (and good spotting!)

File: tests/integration/test_task_pool.py (hunk 13 above)
Comment (oliver-sanders) [reply tests/integration/test_task_pool.py:1756]: Done.

File: tests/integration/scripts/test_validate.py (hunk 12 above)
Comment (oliver-sanders) [reply tests/integration/scripts/test_validate.py]: Done.

File: tests/functional/xtriggers/04-suite_state.t (hunk 11 above)
Comment (oliver-sanders) [reply tests/functional/xtriggers/04-suite_state.t:43]: Reinstated.

File: tests/functional/cylc-install/02-failures.t
Comment (oliver-sanders) [reply tests/functional/cylc-install/02-failures.t:1]: Done.

File: cylc/flow/config.py (hunk 9 above)
Comment (oliver-sanders) [reply cylc/flow/config.py]: Done.

File: cylc/flow/workflow_files.py (hunk 8 above)
Comment (oliver-sanders) [reply cylc/flow/workflow_files.py]: Done.

**Comment by oliver-sanders** [comment]
Rebased, deconflicted and addressed feedback.

**Comment by oliver-sanders** [comment]
Have now added: [9be3f5d](https://github.com/cylc/cylc-flow/pull/7325/commits/9be3f5d1a1e6ef7623665b5bfe0897a4fe3288a0)

* Back support Jinja2 interfaces deprecated at 3.0 and removed at 3.1.
* Jinja2 renamed a bunch of functions and deprecated their old names.
* Our users didn't spot this as Python warnings are not visible to them.
* As of https://github.com/cylc/cylc-flow/pull/7365 (8.6.6) Python warnings
  originating from Jinja2 will be turned into Cylc warnings.
* To provide a migration window, the old function names will be supported
  in Cylc 8.7.x and removed in 8.8.0.

**Review by hjoliver — APPROVED** [review]
Reapproving post @MetRonnie 's much more thorough review! (the changes look good)

File: cylc/flow/task_pool.py (hunk 7 above)
Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: (Have changed to match the `min/default` approach in the block above)

File: tests/integration/scripts/test_install.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
```
Comment (MetRonnie) [inline tests/integration/scripts/test_install.py]:
```suggestion
# Copyright (C) Earth Sciences New Zealand & British Crown (Met Office)
# & Contributors.
```

File: cylc/flow/task_pool.py [hunk 16]
```diff
@@ -432,19 +429,7 @@ def compute_runahead(self, force=False) -> bool:
...
-                    cylc.flow.flags.cylc7_back_compat and
-                    all(
-                        itask.state(TASK_STATUS_FAILED)
-                        for itask in itasks
-                    )
-                ):
-                    continue
-                base_point = point
-                break
+            base_point = min(self.get_tasks_by_point(), default=None)
```
Comment (MetRonnie) [inline cylc/flow/task_pool.py]: The keys of `self.get_tasks_by_point()` are the same as `self.active_tasks` but with some processing of the values that is not needed here.
```suggestion
            base_point = min(self.active_tasks.keys(), default=None)
```

File: cylc/flow/workflow_files.py [hunk 17]
```diff
@@ -1075,27 +1053,47 @@ def is_forbidden(flow_file: Path) -> bool:
...
+        err_msg = NO_FLOW_FILE_MSG_DEPR
+    else:
+        err_msg = NO_FLOW_FILE_MSG
+
     if flow_file_path.is_file():
         detect_both_flow_and_suite(Path(path))
         return flow_file_path
     if suite_rc_path.is_file():
+        err_msg += f'\n\n{SUITE_RC_DEPR_MSG}'
+        if not allow_suite_rc and suite_rc_path.is_file():
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py]:
```suggestion
        if not allow_suite_rc:
```

File: tests/functional/xtriggers/04-suite_state.t [hunk 18]
```diff
@@ -44,8 +44,3 @@ grep_ok "$msg" "${TEST_NAME}.stderr"
 
 # Rename flow.cylc to suite.rc:
 mv "${WORKFLOW_RUN_DIR}/flow.cylc" "${WORKFLOW_RUN_DIR}/suite.rc"
```
Comment (MetRonnie) [inline tests/functional/xtriggers/04-suite_state.t]:
```suggestion
```

File: cylc/flow/cfgspec/globalcfg.py [hunk 19]
```diff
@@ -1454,6 +1450,13 @@ def default_for(
                       [[localhost|cylc-server-..]]  # error
                       [[localhost, cylc-server-..]]  # ok
 
+            .. versionchanged:: 8.0.0
+
+                Many of these settings have replaced those of the same name
+                from the old Cylc 7
+                ``suite.rc[runtime][<namespace>][job]/[remote]``
+                and ``global.rc[hosts][<host>]`` sections.
```
Comment (MetRonnie) [inline cylc/flow/cfgspec/globalcfg.py]: (RST's 3-space indentation)
```suggestion
               Many of these settings have replaced those of the same name
               from the old Cylc 7
               ``suite.rc[runtime][<namespace>][job]/[remote]``
               and ``global.rc[hosts][<host>]`` sections.
```

File: cylc/flow/parsec/jinja2support.py [hunk 20]
```diff
@@ -52,6 +52,44 @@
...
+    Jinja2 renamed a bunch of interfaces. Warnings were added in version 3.0.0,
+    support for the old names was removed in 3.1.0, however, our users did not
+    notice these warnings (you have to manually turn them on) so did not take
+    action.
+
+    In https://github.com/cylc/cylc-flow/pull/7365 (8.6.6) we've captured
+    Jinja2 warnings and turned them into Cylc warnings to ensure they cannot be
+    missed.
+
+    To give users a chance to take action, this patch restores the old
```
Comment (MetRonnie) [inline cylc/flow/parsec/jinja2support.py]: Unless this was meant to be exhaustive,
```suggestion
    To give users a chance to take action, this patch restores some of the old
```

Comment (oliver-sanders) [reply cylc/flow/parsec/jinja2support.py]: This comment is scoped to renamed interfaces (not removed or modified interfaces):

From this line above:

>     Jinja2 renamed a bunch of interfaces. Warnings were added in version 3.0.0,


It should restore all of the renamed interfaces.

File: cylc/flow/workflow_files.py (hunk 17 above)
Comment (oliver-sanders) [reply cylc/flow/workflow_files.py]: Annoyingly, while inspecting this code, I realised that we are still permitting this:

```
flow.cylc -> suite.rc
suite.rc             
```

Which currently produces no warnings?!

This was purposeful, but the context is lost to me (we did briefly create such symlinks to indicate back-compat mode was turned on which may be part of it).

File: cylc/flow/task_pool.py (hunk 16 above)
Comment (oliver-sanders) [reply cylc/flow/task_pool.py]: (Off topic), but yes, this is quite inefficient, I'm not sure how it got like this, will change!

Note, this appears to be the only use of `get_tasks_by_point`!

File: cylc/flow/workflow_files.py (hunk 17 above)
Comment (MetRonnie) [reply cylc/flow/workflow_files.py]: My suggestion was just because `suite_rc_path.is_file()` is duplicated here

File: cylc/flow/parsec/jinja2support.py (hunk 20 above)
Comment (MetRonnie) [reply cylc/flow/parsec/jinja2support.py]: I would add a note that this does not cover all the removed deprecations: https://jinja.palletsprojects.com/en/stable/changes/#version-3-1-0

File: cylc/flow/workflow_files.py (hunk 17 above)
Comment (oliver-sanders) [reply cylc/flow/workflow_files.py]: I know, but looking at it I spotted a problem, a further change is coming...

File: cylc/flow/parsec/jinja2support.py (hunk 20 above)
Comment (oliver-sanders) [reply cylc/flow/parsec/jinja2support.py]: Will clarify thus:

```suggestion
    To give users a chance to take action, this patch restores the renamed
```

This comment does not pertain to the *removed* deprecations, only the *renamed* ones.

**Comment by oliver-sanders** [comment]
@MetRonnie, please the [recent force-push](https://github.com/cylc/cylc-flow/compare/9be3f5d1a1e6ef7623665b5bfe0897a4fe3288a0..5680d31e5c2cb4236d4282da2412be62ff8d78e9) which address feedback above:
* Remove the obsolete `TaskPool.get_tasks_by_point` interface.
* Misc minor stuff.
* Add a test for `get_workflow_title` which also ensures this interface continues to work with `suite.rc` files.
* Refactors whole workflow config file detection stuff, the previous approach was hard to follow, but the motive was restoring support for `flow.cylc -> suite.rc` sylinks which we apparently supported before (support I accidentally removed earlier).

<br />

@hjoliver, you may want to look over the last four commits:
* jinja2: back support removed interfaces:
  * Context here https://matrix.to/#/!gMuCeOJBhYqgWDGwnR:matrix.org/$17835256323042zibit:matrix.org?via=matrix.org
  * Icky back-compat support to be removed ASAP in 8.8
* refactor workflow detection logic:
  * Clean up the code which detects suite.rc / flow.cylc files and determines what is and isn't a valid workflow config.
* scan: add test for get_workflow_title:
   * (Unimportant, a Cylc 7 gcylc thing).
* task_pool: remove obsolete get-tasks_by_point interface:
   * Small unrelated efficiency improvement spotted during review.

File: cylc/flow/parsec/jinja2support.py [hunk 21]
```diff
@@ -52,6 +52,46 @@
...
+    for old, new in (
+        ('contextfilter', 'pass_context'),
+        ('contextfunction', 'pass_context'),
+        ('evalcontextfilter', 'pass_eval_context'),
+        ('evalcontextfunction', 'pass_eval_context'),
+        ('environmentfilter', 'pass_environment'),
+        ('environmentfunction', 'pass_environment'),
+    ):
+        setattr(jinja2, old, getattr(jinja2, new))
+        setattr(jinja2.filters, old, getattr(jinja2, new))
```
Comment (MetRonnie) [inline cylc/flow/parsec/jinja2support.py]: I think we need to provide our own warning about these deprecated imports now, as there won't be a Jinja2-provided warning anymore.

File: cylc/flow/workflow_files.py [hunk 22]
```diff
@@ -1028,74 +1000,106 @@ def get_cylc_run_abs_path(path: Union[Path, str]) -> Union[Path, str]:
...
+    #   8.?
+    # REMOVE AT:
+    #   8.8.0
+    if flow_file_path.resolve().name == WorkflowFiles.SUITE_RC:
+        # flow.cylc -> suite.rc (this is permitted for historical reasons)
+        if allow_suite_rc:
+            LOG.warning(
+                f'{WorkflowFiles.FLOW_FILE} is a symlink to'
+                f' {WorkflowFiles.SUITE_RC} which is now deprecated. Please'
+                f' complete the transition to'' {WorkflowFiles.FLOW_FILE}.'
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py]:
```suggestion
                f' complete the transition to {WorkflowFiles.FLOW_FILE}.'
```

File: tests/integration/validate/test_outputs.py [hunk 23]
```diff
@@ -375,3 +361,48 @@ def test_unique_messages(
...
+    (tmp_path / WorkflowFiles.FLOW_FILE).touch()
+    with pytest.raises(
+        WorkflowFilesError, match=f'Both flow.cylc and suite.rc.*{tmp_path}'
+    ):
+        await validate(ValidateOptions(), str(tmp_path))
+
+    # flow.cylc present
+    (tmp_path / WorkflowFiles.SUITE_RC).unlink()
+    with pytest.raises(WorkflowConfigError):
+        await validate(ValidateOptions(), str(tmp_path))
```
Comment (MetRonnie) [inline tests/integration/validate/test_outputs.py:388]: This should not raise an error, as `flow.cylc` only should be ok? (Currently the error is because the file is empty so not a valid config)

File: tests/unit/test_workflow_files.py [hunk 24]
```diff
@@ -481,6 +390,170 @@ def test_check_flow_file_symlink(
...
+        WorkflowFilesError,
+        match=rf'Both {WorkflowFiles.FLOW_FILE} and {WorkflowFiles.SUITE_RC}',
+    ):
+        check_flow_file(tmp_path, allow_suite_rc=False)
+
+    with pytest.raises(
+        WorkflowFilesError,
+        match=rf'Both {WorkflowFiles.FLOW_FILE} and {WorkflowFiles.SUITE_RC}',
+    ):
+        check_flow_file(tmp_path, allow_suite_rc=False)
```
Comment (MetRonnie) [inline tests/unit/test_workflow_files.py]: I assume this was meant, otherwise this is a duplication
```suggestion
        check_flow_file(tmp_path, allow_suite_rc=True)
```

**Review by MetRonnie — COMMENTED** [review]
(Got a test failure)

File: cylc/flow/parsec/jinja2support.py (hunk 21 above)
Comment (oliver-sanders) [reply cylc/flow/parsec/jinja2support.py]: (thought I had tested that but will have a poke)

File: tests/integration/validate/test_outputs.py (hunk 23 above)
Comment (oliver-sanders) [reply tests/integration/validate/test_outputs.py:388]: The previous two examples raise Workflow**Files**Error, this one raises Workflow**Config**Error which demonstrates the `flow.cylc` file is being accepted.

A little confusing admirably, I've added a `match` pattern and a comment to clarify.

**Comment by oliver-sanders** [comment]
(ISC004 lint failure to be fixed on 8.6.x)

**Comment by oliver-sanders** [comment]
[Latest changes](https://github.com/cylc/cylc-flow/compare/5680d31e5c2cb4236d4282da2412be62ff8d78e9..c043031aa22f690283fcd8c5b82ae91afe5a09fc).

* Outstanding feedback addressed.
* Warning mechanism introduced to ensure those leaning on back-supported interfaces get warned appropriately.
* Functional tests fixed (log message format change).

File: cylc/flow/task_pool.py (hunk 16 above)
Comment (hjoliver) [reply cylc/flow/task_pool.py]: Note https://github.com/cylc/cylc-flow/pull/7400 removes the call to `get_tasks_by_point` from the Cylc 8 branch.

**Comment by hjoliver** [comment]
> @hjoliver, you may want to look over the last four commits: ...

LGTM

**Comment by hjoliver** [comment]
N.B. I had also noticed `get_tasks_by_point` in other work: https://github.com/cylc/cylc-flow/pull/7325#discussion_r3670030420

File: tests/integration/validate/test_outputs.py (hunk 23 above)
Comment (MetRonnie) [reply tests/integration/validate/test_outputs.py:388]: I think it would have been preferable to just have a minimal valid config, but whatever

**Review by MetRonnie — APPROVED** [review]
I've locally run the lint tests that were skipped due to the unrelated flake8 issue

**Comment by MetRonnie** [comment]
> LGTM

Taking that as your approval...
